contracts/src/token/erc20/extensions/flashmint.rs

```diff
@@ -0,0 +1,326 @@
+
+    /// Indicate  an  error related to an invalid flash loan receiver.
+    /// The receiver does not implement the required `onFlashLoan` function.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC3156InvalidReceiver(address receiver);
+
+}
+
+/// Extension of [`Erc20`] that allows token holders to destroy both
+/// their own tokens and those that they have an allowance for,
+/// in a way that can be recognized off-chain (via event analysis).
+pub trait IERC3156FlashLender {
+    /// The error type associated to this ERC-20 Burnable trait implementation.
+    type Error: Into<alloc::vec::Vec<u8>>;
+
+    /// Returns the maximum amount of tokens that can be borrowed
+    /// from this contract in a flash loan.
+    ///
+    /// For tokens that are not supported, this function returns
+    /// `U256::MIN`.
+    ///
+    /// * `token` - The address of the ERC-20 token that will be loaned.
+    fn max_flash_loan(&self, token: Address) -> U256;
+
+    /// Calculates the fee for a flash loan.
+    ///
+    /// The fee is a fixed percentage of the borrowed amount.
+    ///
+    /// If the token is not supported, the function returns an
+    /// `UnsupportedToken` error.
+    ///
+    /// * `token` - The address of the ERC-20 token that will be loaned.
+    /// * `amount` - The amount of tokens that will be loaned.
+    fn flash_fee(
+        &self,
+        token: Address,
+        amount: U256,
+    ) -> Result<U256, Self::Error>;
+
+    /// Executes a flash loan.
+    ///
+    /// This function is part of the ERC-3156 (Flash Loans) standard.
+    ///
+    /// * `receiver` - The contract that will receive the flash loan.
+    /// * `token` - The ERC-20 token that will be loaned.
+    /// * `amount` - The amount of tokens that will be loaned.
+    /// * `data` - Arbitrary data that can be passed to the receiver contract.
+    ///
+    /// The function must return `true` if the flash loan was successful,
+    /// and revert otherwise.
+    fn flash_loan(
+        &mut self,
+        receiver: Address,
+        token: Address,
+        amount: U256,
+        data: AbiBytes,
+    ) -> Result<bool, Self::Error>;
+}
+
+/// A Permit error.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// Indicate an error related to an unsupported loan token.
+    /// This occurs when the specified token cannot be used for loans.
+    UnsupportedToken(ERC3156UnsupportedToken),
+
+    /// Indicate an error related to the loan amount exceeds the maximum.
+    /// The requested amount is higher than the allowed loan for this token
+    /// max_loan.
+    ExceededMaxLoan(ERC3156ExceededMaxLoan),
+
+    /// Indicate  an  error related to an invalid flash loan receiver.
+    /// The receiver does not implement the required `onFlashLoan` function.
+    InvalidReceiver(ERC3156InvalidReceiver),
+
+    /// Error type from [`Erc20`] contract [`erc20::Error`].
+    Erc20(erc20::Error),
+}
+
+sol_storage! {
+    pub struct Erc20Flashmint  {
```

Comment:
it does implement the internal functions _flash_fee and _flash_fee_receiver so they can be overridden and support extensibility, any suggestion of an approach I can use for that

Reply:
The reaseon why i  went with that current approach is because of those internal functions

Reply:
`_flash_fee` and `_flash_fee_receiver` are not part of the spec (see [EIP-3156](https://eips.ethereum.org/EIPS/eip-3156)), so they are not required to exist.
Also, since inheritance is still limited in Stylus, [overriding does not work as you might expect](https://github.com/OffchainLabs/stylus-sdk-rs/issues/109) in other more object-oriented languages.

In short, `_flash_fee` and `_flash_fee_receiver` don't make as much sense in Stylus as they do in Solidity.